torch/_meta_registrations.py

```diff
@@ -2076,22 +2077,96 @@ def meta__scaled_dot_product_flash(
         dtype=torch.float,
         device=query.device,
     )
-    is_sm80 = torch.cuda.is_available() and torch.cuda.get_device_capability() >= (8, 0)
-    is_sm75 = torch.cuda.is_available() and torch.cuda.get_device_capability() >= (7, 5)
-    head_size_rounded = 64 if head_dim <= 64 else 128
-    blocksize_c = (
-        128
-        if (head_size_rounded == 128 and (dropout_p != 0.0 or not is_sm80))
-        or (is_sm75 and head_size_rounded == 64 and dropout_p != 0.0)
-        else 256
-    )
-    max_seqlen_k = math.ceil(max_seqlen_batch_k / blocksize_c) * blocksize_c
-    if max_seqlen_k <= 128:
-        max_seqlen_k = 128
-    elif max_seqlen_k <= 256:
-        max_seqlen_k = 256
-
-    return ouput, logsumexp
+    cumulative_sequence_length_q = torch.empty(
+        batch_size + 1, dtype=torch.int32, device="meta"
+    )
+    cumulative_sequence_length_k = torch.empty(
+        batch_size + 1, dtype=torch.int32, device="meta"
+    )
+
+    if dropout_p > 0.0:
+        DEFAULT_RNG_STATE_SIZE = 816
+        # [Note] This number feels magical but it is derived from the calculation:
+        # pytorch/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:154
+        # where sizeof(4120) will equal 4 bytes for all systems that are
+        # expected to run this meta function :hope:
+        rng_state_tensor = torch.empty(
+            DEFAULT_RNG_STATE_SIZE, dtype=torch.uint8, device="meta"
+        )
+    else:
+        rng_state_tensor = torch.empty(0, dtype=torch.uint8, device="meta")
```

Comment:
@eellison This meta function returns alot of tensors. Most are on GPU but this one in particular is always going to be on CPU. I am failing with:
```
 test/test_ops.py::TestFakeTensorCUDA::test_fake_crossref_backward_amp_nn_functional_scaled_dot_product_attention_cuda_float32 - RuntimeError: Mismatch on aten._scaled_dot_product_flash_attention.default: Devices cpu and cuda:0 are not equal!
 ```
 
 Is there a way to say "meta+cpu" or something? Or shoud I just skip the test?

Reply:
hmm is the first time a pattern like this has come up. the "public api" right now that would allow you to specify return devices would be to register a decomp for flash attention. but I'm not sure when we would ever want to decompose flash attention so I don't know how helpful that is. 

The following should work (haven't tested though):

```
@register_op_impl(aten._sparse_coo_tensor_with_dims_and_tensors.default)
def _flash_attention_backward(fake_mode, func, *args, **kwargs):
    # backward should never use kwargs, right ?
    device_input = args[0].fake_device

    with in_kernel_invocation_manager(fake_mode):
        outputs = func(*args, **kwargs)

    return [
        FakeTensor(fake_mode, out, device_input) for i, out in enumerate(outputs)
        if i != len(outputs) -1 
        else  FakeTensor(fake_mode, out, "cpu")
    ]
 ```


Separate from that - why are we returning a cpu tensor ? that won't work with cudagraphs

Reply:
So the output for scaled_dot_product_attention_forward is -> 
Input tensors are  q, k, v all Cuda tensors
```
(Tensor ouput -> q.options, 
Tensor logsumexp -> q.options, 
Tensor cum_seq_q -> device="cuda", dtype = "long",
Tensor cum_seq_k -> device="cuda", dtype = "long", 
int max_q, int max_k, 
Tensor rng_state -> device="cpu", dtype="uint8"
Tensor debug_attn_mask -> q.options()
```
everything after output is only needed for computing the backward and is used as input for flash_atten_backward. The rng_state thing is interesting. But it should always be CPU bits. I left some comments in the meta_registrations about it.

Reply:
cc @cpuhrsch cpu output will make this not work with cudagraphs

Reply:
cudagraphs (in eager) have a special way of handling random state so that updates are properly done on the GPU, and cudagraphs don't work with non-default generator so that they can always do this special handling.

Reply:
I discussed this with @drisspg. For context, the rng state is saved here as a cpu tensor and then used in the backward to avoid materializing the dropout mask.

We were thinking that the forward kernel should write out the rng state as a cuda tensor. This should work transparently on top of the existing cudagraph rng state handling.

Reply:
After some messages with Natalia I think it should be possible to fully checkpoint the state for backwards with two numbers similar to how Dropout does it if we are using the default generator.